feature: adding `getmail` as an alternative to `fetchmail` (#2803)
docker-mailserver/docker-mailserver: docs/content/config/advanced/mail-getmail.md

---
title: 'Advanced | Email Gathering with Getmail'
---

> To enable the [getmail][getmail-website] service to retrieve e-mails set the environment variable `ENABLE_GETMAIL` to `1`. Your `docker-compose.yml` file should include the following:

Reviewer: ```suggestion
To enable the [getmail][getmail-website] service to retrieve e-mails set the environment variable `ENABLE_GETMAIL` to `1`. Your `compose.yaml` file should include the following:
```

```yaml
environment:
  - ENABLE_GETMAIL=1
  - GETMAIL_POLL=5
```

In your DMS config volume (eg: `docker-data/dms/config/`), create a `getmail-<ID>.cf` file for each remote account that you want to retrieve mail and store into a local DMS account. `<ID>` should be replaced by you, and is just the rest of the filename (eg: `getmail-example.cf`). The contents of each file should be configuration like documented below.

> Your `docker-mailserver` folder should look similar to this example:

Reviewer: ```suggestion
The directory structure should similar to this:
```

```txt
├── docker-data/dms/config
│   ├── dovecot.cf
│   ├── getmail-example.cf
│   ├── postfix-accounts.cf
│   └── postfix-virtual.cf
├── docker-compose.yml
└── README.md
```

## Configuration

A detailed description of the configuration options can be found in the [online version of the manual page][getmail-docs].

### Common Options

> The default options added to each `getmail` config are:

Reviewer: ```suggestion
The default options added to each `getmail` config are:

```

```getmailrc
[options]
verbose = 0
read_all = false
delete = false
max_messages_per_session = 500
received = false
delivered_to = false
```

If you want to use a different base config, mount a file to `/etc/getmailrc_general`. This file will replace the default "Common Options" base config above, that all `getmail-<ID>.cf` files will extend with their configs when used.

> ### IMAP Configuration 
> 
> This example will:
> 
> 1. Connect to the remote IMAP server from Gmail.
> 2. Retrieve mail from the gmail account `alice` with password `notsecure`.
> 3. Store any mail retrieved from the remote mail-server into DMS for the `user1@example.com` account that DMS manages.
> 
> !!! example
>     ```getmailrc
>     [retriever]
>     type = SimpleIMAPRetriever
>     server = imap.gmail.com
>     username = alice
>     password = notsecure
> 
>     [destination]
>     type = MDA_external
>     path = /usr/lib/dovecot/deliver
>     allow_root_commands = true
>     arguments =("-d","user1@example.com")
>     ```

Reviewer: ```suggestion
??? example "IMAP Configuration" 

    This example will:

    1. Connect to the remote IMAP server from Gmail.
    2. Retrieve mail from the gmail account `alice` with password `notsecure`.
    3. Store any mail retrieved from the remote mail-server into DMS for the `user1@example.com` account that DMS manages.

    ```getmailrc
    [retriever]
    type = SimpleIMAPRetriever
    server = imap.gmail.com
    username = alice
    password = notsecure
    [destination]
    type = MDA_external
    path = /usr/lib/dovecot/deliver
    allow_root_commands = true
    arguments =("-d","user1@example.com")
    ```
```

> ### POP3 Configuration
> 
> Just like the IMAP example above, but instead via POP3 protocol if you prefer that over IMAP.
> 
> !!! example
>     ```getmailrc
>     [retriever]
>     type = SimplePOP3Retriever
>     server = pop3.gmail.com
>     username = alice
>     password = notsecure
> 
>     [destination]
>     type = MDA_external
>     path = /usr/lib/dovecot/deliver
>     allow_root_commands = true
>     arguments =("-d","user1@example.com")
>     ```

Reviewer: ```suggestion
??? example "POP3 Configuration"

    Just like the IMAP example above, but instead via POP3 protocol if you prefer that over IMAP.

    ```getmailrc
    [retriever]
    type = SimplePOP3Retriever
    server = pop3.gmail.com
    username = alice
    password = notsecure
    [destination]
    type = MDA_external
    path = /usr/lib/dovecot/deliver
    allow_root_commands = true
    arguments =("-d","user1@example.com")
    ```
```

### Polling Interval

> By default the `getmail` service checks external mail accounts for new mail every 5 minutes. That polling interval is configurable via the `GETMAIL_POLL` ENV variable, with a value in minutes (_default: 5, min: 1, max: 30_):

Reviewer: ```suggestion
By default the `getmail` service checks external mail accounts for new mail every 5 minutes. That polling interval is configurable via the `GETMAIL_POLL` ENV variable, with a value in minutes (_default: 5, min: 1, max: 30_):

```

```yaml
environment:
  - GETMAIL_POLL=1
```

### XOAUTH2 Authentication

> It is possible to utilize the getmail-gmail-xoauth-token helper to provide authentication using xoauth2 for gmail [example 12][getmail-gmail-xoauth-12] or Microsoft Office 365 [example 13][getmail-gmail-xoauth-12]

Reviewer 2: ```suggestion
It is possible to utilize the `getmail-gmail-xoauth-tokens` helper to provide authentication using `xoauth2` for [gmail (example 12)][getmail-docs-xoauth-12] or [Microsoft Office 365 (example 13)][getmail-docs-xoauth-13]
```

[getmail-website]: https://www.getmail.org
[getmail-docs]: https://getmail6.org/configuration.html

> [getmail-gmail-xoauth-12]: https://github.com/getmail6/getmail6/blob/master/docs/getmailrc-examples#L286

Reviewer 2: ```suggestion
[getmail-docs-xoauth-12]: https://github.com/getmail6/getmail6/blob/1f95606156231f1e074ba62a9baa64f892a92ef8/docs/getmailrc-examples#L286
```

> [getmail-gmail-xoauth-12]: https://github.com/getmail6/getmail6/blob/master/docs/getmailrc-examples#L351

Reviewer 2: ```suggestion
[getmail-docs-xoauth-13]: https://github.com/getmail6/getmail6/blob/1f95606156231f1e074ba62a9baa64f892a92ef8/docs/getmailrc-examples#L351
```

```txt
├── docker-data/dms/config
│   ├── dovecot.cf
│   ├── getmail-<ID>.cf
│   ├── postfix-accounts.cf
│   └── postfix-virtual.cf
├── docker-compose.yml
└── README.md
```

## Configuration

A detailed description of the configuration options can be found in the [online version of the manual page][getmail-docs].

### Common Options

The default options added to each `getmail` config are:
```getmailrc
[options]
verbose = 0
read_all = false
delete = false
max_messages_per_session = 500
received = false
delivered_to = false
```

If you want to use different settings mount a filr to /etc/getmailrc_general, this will replace these so you must define all the options you want to use not just the changes.

### IMAP Configuration 

!!! example
    ```getmailrc
    [retriever]
    type = SimpleIMAPRetriever
    server = imap.gmail.com
    username = username
    password = secret

    [destination]
    type = MDA_external
    path = /usr/lib/dovecot/deliver
    allow_root_commands = true
    arguments =("-d","user1@example.com")
    ```

> ### POP3 Configuration 
> 
> !!! example

Reviewer 2: ```suggestion
### POP3 Configuration

Just like the IMAP example above, but instead via POP3 protocol if you prefer that over IMAP.

!!! example
```

    ```getmailrc
    [retriever]
    type = SimplePOP3Retriever
    server = pop3.gmail.com

>     username = username
>     password = secret

Reviewer 2: ```suggestion
    username = alice
    password = notsecure
```

    [destination]
    type = MDA_external
    path = /usr/lib/dovecot/deliver
    allow_root_commands = true

>     arguments =("-d","user2@example.com")

Reviewer 2: Just keeping it consistent with the IMAP config example :)

```suggestion
    arguments =("-d","user1@example.com")
```

    ```

### Polling Interval

By default the `getmail` service checks external mail accounts for new mail every 5 minutes. That polling interval is controlled by the `GETMAIL_POLL` ENV variable, with a value in seconds (default: 5, minimum: 1, maximum: 60):
```yaml
environment:
  - GETMAIL_POLL=1
```

[getmail-website]: https://www.getmail.org
[getmail-docs]: https://getmail6.org/configuration.html

> [getmail-gmail-xoauth]: https://www.bytereef.org/howto/oauth2/getmail.html

Author: yep you are correct, nothing above 30 really makes any sense.
Updated docs to add pointers to the xoauth example configs in the getmail6 github repo.

---
title: 'Advanced | Email Gathering with Getmail'
---

To enable the [getmail][getmail-website] service to retrieve e-mails set the environment variable `ENABLE_GETMAIL` to `1`. Your `docker-compose.yml` file should look like following snippet:

```yaml
environment:
  - ENABLE_GETMAIL=1
  - GETMAIL_POLL=300
```

Generate a file called `getmail-<USER>.cf` and place it in the `docker-data/dms/config/` folder. Your `docker-mailserver` folder should look like this example:

```txt
├── docker-data/dms/config
│   ├── dovecot.cf
│   ├── getmail-<USER>.cf
│   ├── postfix-accounts.cf
│   └── postfix-virtual.cf
├── docker-compose.yml
└── README.md
```

## Configuration

A detailed description of the configuration options can be found in the [online version of the manual page][getmail-docs].

### Common Options

The Default options added to each getmail config are :
```getmailrc
[options]
verbose = 0
read_all = false
delete = false
max_messages_per_session = 500
received = false
delivered_to = false
message_log = /var/log/mail/getmail-PLACEHOLDER.log
```

> to override these mount your own common settings to /etc/getmailrc_general

Author: I've removed the need for the message_log  setting in this file and tried to clarify that mounting a file replaces this one so all options you want need to be defined.

### IMAP Configuration 

!!! example
    ```getmailrc
    [retriever]
    type = SimpleIMAPRetriever
    server = imap.example.net
    username = fred.flintstone
    password = mailpassword

    [destination]
    type = MDA_external
    path = /usr/lib/dovecot/deliver
    allow_root_commands = true
    arguments =("-d","fred.flinstone@bedrock.com")
    ```

### POP3 Configuration 

!!! example
    ```getmailrc
    [retriever]
    type = SimplePOP3Retriever
    server = pop.example.net
    username = fred.flintstone
    password = mailpassword

    [destination]
    type = MDA_external
    path = /usr/lib/dovecot/deliver
    allow_root_commands = true
    arguments =("-d","fred.flinstone@bedrock.com")
    ```

### Polling Interval

By default the getmail service searches every 5 minutes for new mails on your external mail accounts. You can override this default value by changing the ENV variable `GETMAIL_POLL`:

```yaml
environment:
  - GETMAIL_POLL=60
```

> You must specify a numeric argument which is a polling interval in seconds. The example above polls every minute for new mails.

Author: Updated to use minutes, documented minimum and maximum values.
